### PR TITLE
KEYCLOAK-4342 Make adapter state cookie path configurable

### PR DESCRIPTION
**Pull request by scranen**
I proposed this workaround in #4283 to be able to add support for token storage in cookies in the Spring adapter. It touches the `CookieTokenStore` though, which is shared by a number of other adapters.

Key idea was: change the way the cookie path is determined in a backward compatible way, but with the option to override it. The current behaviour assumes that the application is never running on the root path of the server, so we'll need to maintain that as the default. Therefore, I introduced the `adapterStateCookiePath` property in `KeycloakDeployment`, which allows you to configure where the cookie will be stored. If it's a relative path, then it is assumed that the application is running in a context root, and is interpreted relative to that context root. If it's an absolute path, then that path is used to store the cookies. This behaviour is documented by `CookieTokenStoreTest`.

### CODE REVIEW

**Comment by sebastienblanc** [comment]
Could you add a PR to the documentation to describe the `adapterStateCookiePath` property ? 
Besides this , it looks good to me. 
Have you rebased recently ?

**Comment by scranen** [comment]
I did not add the property to the `AdapterConfig`, so this property has the same status as `KeycloakDeployment#setStateCookieName`, which also goes undocumented. I didn't want to add it to the config because I'm not sure whether just adding the field is enough, and whether it will break any APIs (it has Jackson annotations on it so I assume it might be sent over some public API).

I just rebased this branch, and added a commit for something I didn't notice before.

**Comment by scranen** [comment]
If r

**Comment by scranen** [comment]
@sebastienblanc how can we make progress on this PR?

**Comment by stianst** [comment]
I'm afraid we're past feature freeze for Keycloak 3.x and can't consider merging this until we open PRs for 4.x.

Had a quick look and we certainly won't accept this with testing done with Mocks. We have decided to not use Mock libraries so the tests will have to be changed to functional/integration type tests with Arquillian.

**Comment by scranen** [comment]
@stianst Any concrete place you would want to put this test? I don't really see an obvious place to put it, since it's a unit test for a static class that is used by 9 adapters. If we add it to the integration tests, do we need to duplicate this test among the integration tests for all the adapters? Do we choose an arbitrary one?

I would love to contribute, but this is a bit of a demotivating result after 3 months. I suspect I'm going to run into the same issues with #4283, so for that one too: if you could give me any concrete suggestions about what needs to happen in order to get it to an acceptable state, that would be wonderful.

**Comment by stianst** [comment]
I appreciate that, but afraid I can't give you any guidance on how to exactly write the tests as I don't deal with the adapters much at all.

I can also appreciate that it's demotivating and I would love it if we could provide more time to review contributions, but end of the day we have loads on our plate and have to prioritize. If we accept contributions that are not solid or doesn't come with sufficient testing then it ends up being additional work for us.

**Review by pedroigor — CHANGES_REQUESTED** [review]
@scranen, can you rebase this PR, please ?

Also, could you check if org.keycloak.adapters.tomcat.CatalinaHttpFacade.RequestFacade#getRelativePath somehow solves the issue you are trying to fix with your changes ?

I think this method was added after the changes you are proposing and it was intended to help with apps deployed in the root context.

**Comment by pedroigor** [comment]
@scranen, I would say that if this PR does not make sense anymore because of ```getRelativePath``` we should close it.

The ```getRelativePath``` is safe and it was included to solve a similar problem when apps are deployed in the context root.

Were you able to check if your scenario works with current code from upstream ?

**Comment by scranen** [comment]
@pedroigor the issue this PR solves is the incorrect cookie path that is used to set the token cookie. The introduction of `getRelativePath` has not done anything to improve that, since `CookieTokenStore#getContextPath` was not changed (you can see this from the diff). In other words, this PR is still necessary.

To be answered:
1. Does this PR need to preserve backward compatibility? If yes, then deriving the cookie path from `getRelativePath` is not an option, and we should just merge the PR as is.
2. If no, then two questions remain:
2a. Do we even want to derive the cookie path from `getRelativePath`? It doesn't exactly yield clean code; perhaps exposing the context path directly from the HttpFacade.Request would be better (can be done in a separate ticket).
2b. Do we want to keep the cookie path configurable? Since the behaviour is not backward compatible, and some users might rely on the old behaviour, configurability makes it easy to restore the old behaviour.

**Comment by pedroigor** [comment]
@scranen, you are right. I've not considered backward compatibility. In this case, change to ```getRelativePath``` is not an option as we need to keep current behavior. Let`s keep as it is.

For last, I would ask you to remove mockito from this PR (and tests) and implement a test method in this base class ```org.keycloak.testsuite.adapter.servlet.AbstractDemoServletsAdapterTest```. In our testsuite, you'll see that tests extending this base class are running for different containers such as Wildfly and Undertow.

To run this test you use a command as follows:

```bash
mvn -f testsuite/integration-arquillian/pom.xml clean install -Dtest=UndertowDemoServletsAdapterTest
```

There are some methods in that base class that is checking the cookie.

If you are not able to write the test by your own I can cherry-pick your changes and write one.

**Comment by scranen** [comment]
@pedroigor can you give me some pointers on how to change the context root that the CustomerCookiePortal gets deployed on? The only thing I could find about changing the context root in Arquillian was to define it as a `contextRoot` property in the protocol in `arquillian.xml`, but that doesn't seem to have any effect.

**Comment by pedroigor** [comment]
@scranen, can you check if this works for you ?

```code
       WebArchive original = exampleDeployment(RESOURCE_SERVER_ID);

        WebArchive archive = ShrinkWrap.create(WebArchive.class, "ROOT.war");

        archive.merge(original);

        return archive;
```
Use this code within the method that creates de deployment archive.

**Comment by pedroigor** [comment]
@scranen, don't you also need to change ```AdapterConfig``` to parse this new property from ```keycloak.json``` (adapter config file) ?

**Comment by scranen** [comment]
@pedroigor thanks, the merge trick works to a point (it seems to unexpectedly deploy `ROOT.war` to context root `/ROOT`, but I can trick it by using `.war` as a file name. Very dubious, so if you know of a better way...?).

I bumped into a few more issues because the context root is now sometimes empty, but the browser automatically converts that to path `/`, so I had to make some changes to deal with that.

Initially I opted for not parsing this property from `keycloak.json` (see my second comment in this thread). I can see that it's easier to have it there for testing purposes, so I'll add it after all. I will try to get a new commit in by tomorrow night.

**Comment by scranen** [comment]
Added a commit. I noticed the merge conflicts with the upstream, but it seems that the command you gave to run this particular test is no longer working after a rebase, and a full build was giving me errors in what seemed to be an unrelated test. Will look at this again when I have some more time to spend on this. Any pointers would again be very welcome.

**Comment by pedroigor** [comment]
@scranen, thanks for all the work. The new command to run tests is:

```bash
mvn -Pauth-server-wildfly -Papp-server-wildfly clean install -Djs.browser=firefox -Dfirefox_binary=/pedroigor/development/tools/firefox/firefox -Dsurefire.skipAfterFailureCount=1 -Dtest=DemoServletsAdapterTest
```

**Comment by scranen** [comment]
@pedroigor I had to run a maven install with `-Pdistribution` first before I could run that command. Got it to run now, but the test is failing for me with the following error:
```
16:30:08,859 ERROR [org.keycloak.testsuite.adapter.servlet.DemoServletsAdapterTest] [DemoServletsAdapterTest] null() FAILED
Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 91.465 sec <<< FAILURE! - in org.keycloak.testsuite.adapter.servlet.DemoServletsAdapterTest
org.keycloak.testsuite.adapter.servlet.DemoServletsAdapterTest  Time elapsed: 91.399 sec  <<< ERROR!
java.lang.RuntimeException: Could not create statement
Caused by: org.jboss.arquillian.drone.impl.DroneTimeoutException: Drone creation request timed out after 60 seconds. Make sure that browser or remote server in case of remotely executed driver is running and communication with Drone haven't failed.
Caused by: java.lang.RuntimeException: 
Unable to instantiate Drone via org.openqa.selenium.firefox.FirefoxDriver(FirefoxOptions): org.openqa.selenium.WebDriverException: Timed out waiting 45 seconds for Firefox to start.
Build info: version: '3.11.0', revision: 'e59cfb3', time: '2018-03-11T20:26:55.152Z'
System info: host: 'shunkle.local', ip: '192.168.44.129', os.name: 'Mac OS X', os.arch: 'x86_64', os.version: '10.13.4', java.version: '1.8.0_131'
Driver info: driver.version: FirefoxDriver
Caused by: org.openqa.selenium.WebDriverException: 
Timed out waiting 45 seconds for Firefox to start.
Build info: version: '3.11.0', revision: 'e59cfb3', time: '2018-03-11T20:26:55.152Z'
System info: host: 'shunkle.local', ip: '192.168.44.129', os.name: 'Mac OS X', os.arch: 'x86_64', os.version: '10.13.4', java.version: '1.8.0_131'
Driver info: driver.version: FirefoxDriver
Caused by: org.openqa.selenium.net.UrlChecker$TimeoutException: Timed out waiting for [http://localhost:44010/hub/status] to be available after 45002 ms
Caused by: java.util.concurrent.TimeoutException
```
I can see that it's started Firefox successfully, but apparently it's not able to connect. Not sure how to proceed, should I just push the rebased branch?

**Comment by scranen** [comment]
Okay, left out the firefox parameters, now the test is running again, but the trick to get it deployed on `/` seems to be broken now.

**Comment by scranen** [comment]
@pedroigor using a `ROOT.war` again makes it deploy to context root `/ROOT`, and my trick to call the archive `.war` no longer works because after rebasing it does not accept an empty filename anymore. Shall I push the rebased version for you to have a go?

**Comment by pedroigor** [comment]
@scranen, sorry for the late response. 

Yes, please. Rebase this and I'll try to make your test running fine.

**Comment by scranen** [comment]
Rebased.

**Comment by pedroigor** [comment]
@scranen, tks. Let's see how tests goes now ...

**Comment by pedroigor** [comment]
@scranen, could you please take a look at this commit https://github.com/pedroigor/keycloak/commit/69c6f2cab7e52e157110fd0c2b10726fbde97052 ?

I have created a ```CookieStoreRootContextTest``` using ```token-cookie-path```. It is running fine, feel free to add more tests if you find necessary. Maybe we need to also test relative paths ?

**Comment by fluescher** [comment]
Thank you all for you work. Can I help somehow in getting this merged?

**Comment by pedroigor** [comment]
@fluescher, in my last comment I mentioned some changes I did here https://github.com/pedroigor/keycloak/commit/69c6f2cab7e52e157110fd0c2b10726fbde97052. They are basically adding an arquillian test and also changing adapter config to include a `token-cookie-path` option. Are you interested in  taking this and include tests + new option ?

**Comment by fluescher** [comment]
@pedroigor I really like to use #4283 so I am happy to support. I'm unsure of what still need to be done. What exactly is missing? Thanks for your advice.

In the meantime I read trought the history of this PR.

**Comment by vsomasvr** [comment]
@pedroigor and @scranen Thank you for all great work.
Can any of you please specify what exactly is blocking this PR, so that its clear for anybody planning to contribute.
This is specified as a pre-req for #4283 (KEYCLOAK-5130), and sorry to see a critical feature not moving forward from PR queue.

**Comment by scranen** [comment]
Sorry for not checking in for so long. @pedroigor created tests, I added his commit to this branch. As you can see, the tests failed in the automated build, but I have not been able to reproduce the failure locally. My understanding is that once the tests compile, the PR is good to merge (but in the end that is not my call).

**Comment by pedroigor** [comment]
@scranen, will check if it works for me too and see why it is failing in travis. Thanks for continuing this.

**Comment by fluescher** [comment]
@pedroigor & @scranen : I was able to reproduce it locally. It seems the archive was not deployed to /

Please have a look at commit fluescher@42be21511069a3e64f518f4aab2213947fe16b0b These changes make the test pass locally.

**Comment by pedroigor** [comment]
@fluescher, I was not able to run the test locally, It seems arquillian is complaining because you just left `.war`.

@scranen, I run the test again locally and it works, I would like to restart a build but we need to rebase your branch first to solve some compilation problems. If you can do this, I'll wait. However, I'm also fine to close this PR and submit a new one with your commits + any other I may need to get this in.

**Comment by scranen** [comment]
Rebased on upstream master

**Comment by scranen** [comment]
Got it to compile again, but tests are failing for me:
```
Failed tests: 
  JavascriptAdapterTest.fragmentInLoginFunction:564->assertOnTestAppUrl:102 URL expected to begin with:http://localhost:8180/auth/realms/test/testing/javascript/index.html; actual URL: http://localhost:8180/auth/realms/test/login-actions/required-action?execution=OAUTH_GRANT&client_id=js-console&tab_id=KgS1G5VJHJs
  JavascriptAdapterTest.fragmentInURLTest:551->assertOnTestAppUrl:102 URL expected to begin with:http://localhost:8180/auth/realms/test/testing/javascript/index.html; actual URL: http://localhost:8180/auth/realms/test/login-actions/required-action?execution=OAUTH_GRANT&client_id=js-console&tab_id=EL_MOLBdo_I
  JavascriptAdapterTest.implicitFlowTest:229->assertOnTestAppUrl:102 URL expected to begin with:http://localhost:8180/auth/realms/test/testing/javascript/index.html; actual URL: http://localhost:8180/auth/realms/test/login-actions/required-action?execution=OAUTH_GRANT&client_id=js-console&tab_id=uHWCCguD_rk
  JavascriptAdapterTest.loginRequiredAction:326->assertOnTestAppUrl:102 URL expected to begin with:http://localhost:8180/auth/realms/test/testing/javascript/index.html; actual URL: http://localhost:8180/auth/realms/test/login-actions/required-action?execution=OAUTH_GRANT&client_id=js-console&tab_id=Ma-7rg9Hcfk
  JavascriptAdapterTest.testGetProfile:164->assertOnTestAppUrl:102 URL expected to begin with:http://localhost:8180/auth/realms/test/testing/javascript/index.html; actual URL: http://localhost:8180/auth/realms/test/login-actions/required-action?execution=OAUTH_GRANT&client_id=js-console&tab_id=CskWrsIufrQ
  JavascriptAdapterTest.testLoginWithKCLocale:131->assertOnTestAppUrl:102 URL expected to begin with:http://localhost:8180/auth/realms/test/testing/javascript/index.html; actual URL: http://localhost:8180/auth/realms/test/login-actions/required-action?execution=OAUTH_GRANT&client_id=js-console&tab_id=KvYYJ-QZdH4
  JavascriptAdapterTest.testRefreshToken:143->assertOnTestAppUrl:102 URL expected to begin with:http://localhost:8180/auth/realms/test/testing/javascript/index.html; actual URL: http://localhost:8180/auth/realms/test/login-actions/required-action?execution=OAUTH_GRANT&client_id=js-console&tab_id=AF8WN3QqpjE
  JavascriptAdapterTest.testRefreshTokenIfUnder30s:152->assertOnTestAppUrl:102 URL expected to begin with:http://localhost:8180/auth/realms/test/testing/javascript/index.html; actual URL: http://localhost:8180/auth/realms/test/login-actions/required-action?execution=OAUTH_GRANT&client_id=js-console&tab_id=3S5jJnEtNyo
  CookieStoreRootContextTest.testTokenInCookieSSORoot:95->loginToCustomerCookiePortalRoot:126 URL expected to begin with:http://localhost:8180/auth/realms/demo/protocol/openid-connect/auth; actual URL: http://localhost:8280/
Tests in error: 
  ServletAuthzNoLazyLoadPathsAdapterTest>AbstractServletAuthzAdapterTest.testGrantPremiumAccessToUser:141->AbstractBaseServletAuthzAdapterTest.performTests:68->AbstractBaseServletAuthzAdapterTest.performTests:79 » Runtime
  JavascriptAdapterTest.grantBrowserBasedApp:206->lambda$grantBrowserBasedApp$99c597a6$2:206 » Runtime
  JavascriptAdapterTest.implicitFlowCertEndpoint:283 » WebDriver {"errorMessage"...
  JavascriptAdapterTest.implicitFlowOnTokenExpireTest:270 » WebDriver {"errorMes...
  JavascriptAdapterTest.implicitFlowRefreshTokenTest:258 » WebDriver {"errorMess...
  JavascriptAdapterTest.initializeWithRefreshToken:515 » WebDriver {"errorMessag...
  JavascriptAdapterTest.initializeWithTimeSkew:480 » WebDriver {"errorMessage":"...
  JavascriptAdapterTest.initializeWithTokenTest:453 » WebDriver {"errorMessage":...
  JavascriptAdapterTest.reentrancyCallbackTest:527 » WebDriver {"errorMessage":"...
  JavascriptAdapterTest.testBearerRequest:310 » WebDriver {"errorMessage":"Can't...
  JavascriptAdapterTest.testCertEndpoint:236 » WebDriver {"errorMessage":"Can't ...
  EntitlementAPITest.testOfflineRequestingPartyToken:1202 » Runtime Error execut...
  ResetPasswordTest.resetPassword » Timeout java.net.SocketTimeoutException: Rea...
  ResetPasswordTest.resetPasswordByEmail » Timeout java.net.SocketTimeoutExcepti...
  ResetPasswordTest.resetPasswordCancelChangeUser » Timeout java.net.SocketTimeo...
  ResetPasswordTest.resetPasswordExpiredCode » Timeout java.net.SocketTimeoutExc...
  ResetPasswordTest.resetPasswordExpiredCodeAndAuthSession » Timeout java.net.So...
  ResetPasswordTest.resetPasswordExpiredCodeAndAuthSessionPerActionLifespan » Timeout
  ResetPasswordTest.resetPasswordExpiredCodeAndAuthSessionPerActionMultipleTimeouts » Timeout
  ResetPasswordTest.resetPasswordExpiredCodeForgotPasswordFlow » Timeout java.ne...
  ResetPasswordTest.resetPasswordExpiredCodeForgotPasswordFlowPerActionLifespan » Timeout
  ResetPasswordTest.resetPasswordExpiredCodeForgotPasswordFlowPerActionMultipleTimeouts » Timeout
  ResetPasswordTest.resetPasswordExpiredCodeShort » Timeout java.net.SocketTimeo...
  ResetPasswordTest.resetPasswordExpiredCodeShortPerActionLifespan » Timeout jav...
  ResetPasswordTest.resetPasswordExpiredCodeShortPerActionMultipleTimeouts » Timeout
  ResetPasswordTest.resetPasswordLink » Timeout java.net.SocketTimeoutException:...
  ResetPasswordTest.resetPasswordLinkNewBrowserSessionPreserveClient » Timeout j...
  ResetPasswordTest.resetPasswordLinkOpenedInNewBrowser » Timeout java.net.Socke...
  ResetPasswordTest.resetPasswordLinkOpenedInNewBrowserAndAccountClientRenamed » Timeout
  ResetPasswordTest.resetPasswordTwice » Timeout java.net.SocketTimeoutException...
  ResetPasswordTest.resetPasswordTwiceInNewBrowser » Timeout java.net.SocketTime...
  ResetPasswordTest.resetPasswordWithLengthPasswordPolicy » Timeout java.net.Soc...
  ResetPasswordTest.resetPasswordWithPasswordHistoryPolicy » Timeout java.net.So...
  ResetPasswordTest.resetPasswordWithSpacesInUsername » Timeout java.net.SocketT...

Tests run: 2106, Failures: 9, Errors: 34, Skipped: 200
```

**Comment by pedroigor** [comment]
@scranen, same error. Will try to figure out the cause tomorrow.

**Comment by fluescher** [comment]
@pedroigor I was able to fix exactly this behaviour by using just the `.war` as name and another change in fluescher/keycloak@42be215

**Comment by pedroigor** [comment]
@scranen, Could you include these two commits in your PR:

* https://github.com/pedroigor/keycloak/commit/8b0a6f61c297dfa8521dae1f912dabfc5f3388ec
* https://github.com/pedroigor/keycloak/commit/0301260a1c840d3175805a792159b146f0ff0436

I think I fixed the issue with that test that is failing. Basically, the first commit above allows deployments to use the root context path when tests are running using Undertow.

Also, in order to finally get this done, please remove all all those tests/references to mockito.

**Comment by pedroigor** [comment]
@scranen, in addition to those two commits, could you also include this one https://github.com/pedroigor/keycloak/commit/d9701916fd1556545bf49115df8f269b07446449. It basically changes the test to better assert the changes we are making here.

Btw, after doing all these things we discussed, could you squash all commits in a single one ?

**Comment by scranen** [comment]
Hi @pedroigor, I made a change to remove Mockito from the tests, and instead use manually written mocks as is done elsewhere in the project (although I should point out that Mockito too is used in the Spring security, Spring Boot and Spring Boot 2 adapters). If you really want the test removed as well, that can be arranged, but it does mean coverage goes down again. Let me know what you want and I'll squash the commits.

**Comment by pedroigor** [comment]
@scranen, you are right about mockito being used in spring adapters. But that is something we want to change in the future with the proper integration tests.

Please, remove that test. I think my latest changes to {{CookieStoreRootContextTest}} are covering the new config option we are discussing and existing tests also make sure we are not breaking anything. However, if you feel we could improve that integration test to improve coverage, you can always add more tests there.

I think we are finally close to get this one in :) Please, squash and let's wait for CI outcome.

**Comment by pedroigor** [comment]
@scranen, thanks. We are very close to merge this now. Glad that tests are now passing.

FYI, we are doing some changes related with WF14 upgrade and we are in the middle of testing the upcoming changes. We will merge this one right after we have those WF14 changes in.

This PR LGTM now. However, I forgot to include the necessary changes to the subsystem. Sorry about that. Could you please cherry-pick this https://github.com/pedroigor/keycloak/commit/ac454793a2879b8146d3124dd63563011eb407f2 and squash it in this PR?

I will also send a PR to documentation, you don't need to worry about this.

**Review by pedroigor — DISMISSED** [review]

**Comment by scranen** [comment]
Done.

File: adapters/oidc/adapter-core/src/main/java/org/keycloak/adapters/CookieTokenStore.java
Comment (hmlnarik) [inline adapters/oidc/adapter-core/src/main/java/org/keycloak/adapters/CookieTokenStore.java]: This is going to throw NPE when adapterStateCookiePath is not set (even though in this case it is initialized in KeycloakDeployment, sanity checks are generally preferred since this can be changed in the future)

File: adapters/oidc/adapter-core/src/main/java/org/keycloak/adapters/CookieTokenStore.java
Comment (hmlnarik) [inline adapters/oidc/adapter-core/src/main/java/org/keycloak/adapters/CookieTokenStore.java]: deployment.getAdapterStateCookiePath().trim().isEmpty()

**Review by hmlnarik — COMMENTED** [review]
On the first sight, looks reasonable. Few points though:

- `adapterStateCookiePath`, `COOKIE_PATH`, `token-cookie-path`, `keycloak.secure-server.token-cookie-path` etc. refer to the same thing. Why are they named differently? Should be consistent.
- `token-cookie-path` and the its behaviour should be documented before merging, as requested above. The PR can only be merged together with the docs.

**Comment by pedroigor** [comment]
@scranen, I've sent a PR [1] with the documentation changes. Now that Wildfly 14 changes are in, we are ready to merge these changes. I'll start a new build in Jenkins too to make sure tests are running fine after Wildfly 14 changes.

Meanwhile, could you please consider Hynek comments?

[1] https://github.com/keycloak/keycloak-documentation/pull/504

**Comment by scranen** [comment]
I originally called the configuration property `token-cookie-path` because I thought it would be easier to find in the documentation when you look for it (the property that enables the cookie is called `token-store`, not `adapter-state-store`), and I did not think it would be appropriate to rename it in the code in this PR. I have no strong opinion though, so changed it to `adapter-state-cookie-path` now.

**Comment by pedroigor** [comment]
@scranen, what about the other two comments from Hynek ? If you work them, I can work to get this in ASAP :) Finally !

**Comment by pedroigor** [comment]
@scranen, what about this one https://github.com/keycloak/keycloak/pull/4283 ? Most of the changes here are there. We could probably start reviewing that one next week once we get this PR in. Not sure exactly, what is the actuall diff in that PR after the changes in this PR.

**Comment by scranen** [comment]
@pedroigor which other 2 comments? (I addressed 2 comments in the first commit + the naming in the second commit)

**Comment by scranen** [comment]
I've started rebasing the other one btw, just need to find some more time to do it because there are some conflicts.

**Comment by pedroigor** [comment]
@scranen, yeah. I was in an outdated version of this page when making that comment. Please, ignore it. Let's wait for CI now.

**Comment by pedroigor** [comment]
@keycloak-bot test

**Comment by pedroigor** [comment]
@keycloak-bot-ci test

**Review by pedroigor — APPROVED** [review]

**Comment by pedroigor** [comment]
@hmlnarik, could you please review this one again ? @scranen did the changes you suggested.

**Comment by pedroigor** [comment]
@scranen, tks for the long and hard work :)

**Pull request merged**
